File: temp/add_time_field.py
#!/usr/bin/env python3
"""
add_time_field.py — Add a 'time' field to PointCloud2 for FAST-LIO compatibility.

Gazebo's gpu_lidar outputs PointCloud2 with fields (x, y, z, intensity) but
FAST-LIO requires a 'time' field for each point (used in deskewing/motion
compensation). This relay node adds a zero-filled 'time' field.

Usage:
  source install/setup.bash
  python3 temp/add_time_field.py
"""

import rclpy
from rclpy.node import Node
from rclpy.parameter import Parameter
from sensor_msgs.msg import PointCloud2, PointField


class AddTimeField(Node):
    def __init__(self):
        super().__init__("add_time_field",
                         parameter_overrides=[Parameter('use_sim_time', Parameter.Type.BOOL, True)])

        self.declare_parameter("input_topic", "/x500_lidar/scan/points")
        self.declare_parameter("output_topic", "/x500_lidar/scan/points_timed")

        in_topic = self.get_parameter("input_topic").value
        out_topic = self.get_parameter("output_topic").value

        self._sub = self.create_subscription(
            PointCloud2, in_topic, self._cb, rclpy.qos.QoSProfile(depth=5, reliability=rclpy.qos.ReliabilityPolicy.BEST_EFFORT))
        self._pub = self.create_publisher(
            PointCloud2, out_topic, rclpy.qos.QoSProfile(depth=5, reliability=rclpy.qos.ReliabilityPolicy.BEST_EFFORT))

        self.get_logger().info(f"Adding time field: {in_topic} → {out_topic}")

    def _cb(self, cloud: PointCloud2):
        # Check if 'time' field already exists
        field_names = [f.name for f in cloud.fields]
        if 'time' in field_names:
            self._pub.publish(cloud)
            return

        # Compute point count and data layout
        point_step_old = cloud.point_step
        n_points = cloud.width * cloud.height

        # New layout: original fields + time (float32, 4 bytes)
        time_offset = point_step_old
        new_point_step = point_step_old + 4  # add float32 time
        new_row_step = new_point_step * cloud.width
        new_data = bytearray(n_points * new_point_step)

        import struct
        # timestamp_unit=2 (microseconds), scan_period=1/10Hz=0.1s=100,000us
        t_step = 100000.0 / n_points if n_points > 0 else 0.0

        for i in range(n_points):
            src_start = i * point_step_old
            dst_start = i * new_point_step
            new_data[dst_start:dst_start + point_step_old] = \
                cloud.data[src_start:src_start + point_step_old]
            # time field: float32 = point timestamp in microseconds
            t_us = i * t_step
            new_data[dst_start + time_offset:dst_start + new_point_step] = \
                struct.pack('<f', t_us)

        # Build new message
        new_cloud = PointCloud2()
        new_cloud.header = cloud.header
        new_cloud.height = cloud.height
        new_cloud.width = cloud.width
        new_cloud.is_dense = cloud.is_dense
        new_cloud.is_bigendian = cloud.is_bigendian
        new_cloud.point_step = new_point_step
        new_cloud.row_step = new_row_step
        new_cloud.data = bytes(new_data)

        # Copy original fields + add 'time'
        new_cloud.fields = list(cloud.fields)
        time_field = PointField()
        time_field.name = 'time'
        time_field.offset = time_offset
        time_field.datatype = PointField.FLOAT32
        time_field.count = 1
        new_cloud.fields.append(time_field)

        self._pub.publish(new_cloud)


def main():
    rclpy.init()
    rclpy.spin(AddTimeField())
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/lidar_bridge/src/add_time_field.cpp
// add_time_field — Add a 'time' field to PointCloud2 for FAST-LIO and
// monotonicize the scan header stamp.
//
// Gazebo's gpu_lidar outputs PointCloud2 with fields (x, y, z, intensity) but
// FAST-LIO requires a 'time' field per point (used in deskewing). This node
// appends a zero-filled float32 'time' field. It also clamps the forwarded
// header stamp to a strictly increasing sequence (see stamp_util.hpp) so
// FAST-LIO never sees a regressing scan time.
//
// Usage:
//   ros2 run lidar_bridge add_time_field --ros-args \
//     -p input_topic:=/swan_gamma_v2/scan_left/points \
//     -p output_topic:=/swan_gamma_v2/scan_left/points_timed
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/point_cloud2.hpp>
#include <sensor_msgs/msg/point_field.hpp>

#include <algorithm>
#include <cstdint>
#include <cstring>
#include <string>
#include <vector>

#include "lidar_bridge/stamp_util.hpp"

using sensor_msgs::msg::PointCloud2;
using sensor_msgs::msg::PointField;

namespace lidar_bridge
{

class AddTimeField : public rclcpp::Node
{
public:
  AddTimeField() : Node("add_time_field")
  {
    declare_parameter("input_topic", "/x500_lidar/scan/points");
    declare_parameter("output_topic", "/x500_lidar/scan/points_timed");

    const std::string in_topic = get_parameter("input_topic").as_string();
    const std::string out_topic = get_parameter("output_topic").as_string();

    auto qos = rclcpp::QoS(5).best_effort();
    sub_ = create_subscription<PointCloud2>(
        in_topic, qos,
        [this](const PointCloud2::SharedPtr cloud) { onCloud(cloud); });
    pub_ = create_publisher<PointCloud2>(out_topic, qos);

    RCLCPP_INFO(get_logger(), "add_time_field: %s -> %s", in_topic.c_str(),
                out_topic.c_str());
  }

private:
  void onCloud(const PointCloud2::SharedPtr cloud)
  {
    ++cb_count_;
    auto out = std::make_shared<PointCloud2>();
    *out = *cloud;
    out->header.stamp = mono_.clamp(out->header.stamp);

    const bool has_time = std::any_of(
        cloud->fields.begin(), cloud->fields.end(),
        [](const PointField &f) { return f.name == "time"; });

    if (!has_time) {
      // Append a zero-filled float32 'time' field at the end of each point.
      const std::uint32_t point_step = cloud->point_step;
      const std::uint32_t n = cloud->width * cloud->height;
      out->point_step = point_step + 4;
      out->row_step = out->point_step * cloud->width;
      out->data.resize(static_cast<std::size_t>(n) * out->point_step);

      // Copy existing bytes, then zero the tail (the new time field).
      for (std::uint32_t i = 0; i < n; ++i) {
        const std::size_t src = static_cast<std::size_t>(i) * point_step;
        const std::size_t dst = static_cast<std::size_t>(i) * out->point_step;
        std::memcpy(out->data.data() + dst, cloud->data.data() + src, point_step);
        std::memset(out->data.data() + dst + point_step, 0, 4);
      }

      PointField time_field;
      time_field.name = "time";
      time_field.offset = point_step;
      time_field.datatype = PointField::FLOAT32;
      time_field.count = 1;
      out->fields.push_back(time_field);
    }

    maybeLogWarning();
    pub_->publish(*out);
  }

  void maybeLogWarning()
  {
    if (mono_.clamped() > 0 && cb_count_ % 500 == 0) {
      RCLCPP_WARN(get_logger(), "add_time_field: %lu/%lu stamps clamped "
                                "(non-monotonic sim clock)",
                  static_cast<unsigned long>(mono_.clamped()),
                  static_cast<unsigned long>(cb_count_));
    }
  }

  rclcpp::Subscription<PointCloud2>::SharedPtr sub_;
  rclcpp::Publisher<PointCloud2>::SharedPtr pub_;
  StampMonotonicizer mono_;
  std::uint64_t cb_count_{0};
};

}  // namespace lidar_bridge

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<lidar_bridge::AddTimeField>());
  rclcpp::shutdown();
  return 0;
}
